drivers/wireless/lpwan/rn2xx3/rn2xx3.c

```diff
@@ -0,0 +1,1564 @@
+
+static int rn2xx3_gettxpwr(FAR struct rn2xx3_dev_s *priv, FAR int8_t *txpwr)
+{
+  ssize_t length;
+  char response[20];
+
+  length =
+      rn2xx3_getparam(priv, "radio get pwr\r\n", response, sizeof(response));
+  if (length < 0)
+    {
+      return length;
+    }
+
+  *txpwr = atoi(response);
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rn2xx3_setbw
+ *
+ * Description:
+ *    Set the bandwidth of the RN2xx3.
+ *
+ ****************************************************************************/
+
+static int rn2xx3_setbw(FAR struct rn2xx3_dev_s *priv, uint32_t bw)
+{
+  char bwstr[16];
+  ssize_t length;
+  length =
+      file_write(&priv->uart, "radio set bw ", sizeof("radio set bw ") - 1);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  /* Write actual parameter and end command */
+
+  length = snprintf(bwstr, sizeof(bwstr), "%lu\r\n", bw);
+  length = file_write(&priv->uart, bwstr, length);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  return get_command_err(priv);
+}
+
+/****************************************************************************
+ * Name: rn2xx3_getbw
+ *
+ * Description:
+ *    Get the bandwidth of the RN2xx3.
+ *
+ ****************************************************************************/
+
+static int rn2xx3_getbw(FAR struct rn2xx3_dev_s *priv, FAR uint32_t *bw)
+{
+  int length;
+  char response[10];
+
+  length =
+      rn2xx3_getparam(priv, "radio get bw\r\n", response, sizeof(response));
+  if (length < 0)
+    {
+      return length;
+    }
+
+  *bw = strtoul(response, NULL, 10);
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rn2xx3_setsf
+ *
+ * Description:
+ *    Set the spread factor of the RN2xx3.
+ *
+ ****************************************************************************/
+
+static int rn2xx3_setsf(FAR struct rn2xx3_dev_s *priv, uint8_t sf)
+{
+  char sfstr[15];
+  ssize_t length;
+
+  length =
+      file_write(&priv->uart, "radio set sf ", sizeof("radio set sf ") - 1);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  /* Write actual parameter and end command */
+
+  length = snprintf(sfstr, sizeof(sfstr), "sf%u\r\n", sf);
+  length = file_write(&priv->uart, sfstr, length);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  return get_command_err(priv);
+}
+
+/****************************************************************************
+ * Name: rn2xx3_getsf
+ *
+ * Description:
+ *    Get the spread factor of the RN2xx3.
+ *
+ ****************************************************************************/
+
+static int rn2xx3_getsf(FAR struct rn2xx3_dev_s *priv, FAR uint8_t *sf)
+{
+  ssize_t length;
+  char response[20];
+
+  length =
+      rn2xx3_getparam(priv, "radio get sf\r\n", response, sizeof(response));
+  if (length < 0)
+    {
+      return length;
+    }
+
+  *sf = strtoul(&response[2], NULL, 10); /* Skip 'sf' */
```

Comment:
All these functions lack any data validation. Sooner or later (read, within first week after prod deployment) you will start sending invalid data to userspace, not even invalid but just garbage.

Userspace should trust data that comes in from kernel, kernel should not trust data from hardware and should validate it as much as it can.

You should at least check if read data is within parameters of what hardware is expected to send